typed-racket-lib/typed-racket/types/subtype.rkt

```diff
@@ -221,532 +297,463 @@
+  (-> (listof (cons/c Type? Type?)) Result? Result?
+        any/c)
+  (match* (res1 res2)
+    [((Result: t1 (PropSet: p1+ p1-) o1)
+      (Result: t2 (PropSet: p2+ p2-) o2))
+     (and (or (equal? o1 o2) (Empty? o2) (not o2))
+          (subtype-seq A
+                       (subtype* t1 t2)
+                       (prop-subtype* p1+ p2+)
+                       (prop-subtype* p1- p2-)))]))
+
+;;************************************************************
+;; Type Subtyping
+;;************************************************************
+
+(define/cond-contract (type-equiv? A t1 t2)
+  (-> list? Type? Type? any/c)
+  (subtype-seq A
+	       (subtype* t1 t2)
+	       (subtype* t2 t1)))
 
 ;; the algorithm for recursive types transcribed directly from TAPL, pg 305
 ;; List[(cons Number Number)] type type -> List[(cons Number Number)] or #f
 ;; is s a subtype of t, taking into account previously seen pairs A
-(define/cond-contract (subtype* A s t)
-  (c:-> (c:listof (c:cons/c fixnum? fixnum?)) Type? Type? c:any/c)
-  (define ss (Rep-seq s))
-  (define st (Rep-seq t))
+;;
+;; <><NOTE><> the seen list (A) should be updated for the following
+;; types as they are encountered:
+;; needs-resolved? types (Mus, Names, Apps),
+;; Instances, and Structs (Prefabs?)
+(define/cond-contract (subtype* A t1 t2)
+  (-> (listof (cons/c Type? Type?)) Type? Type? (or/c #f list?))
   (early-return
-   #:return-when (or (eq? ss st) (seen? ss st A)) A
-   (define cr (let ([inner (hash-ref subtype-cache st #f)])
-                (if inner
-                    (hash-ref inner ss 'missing)
-                    'missing)))
+   #:return-when (seen? t1 t2 A) A
+   #:return-when (Univ? t2) A
+   ;; error is top and bot
+   #:return-when (or (type-equal? t1 Err)
+                     (type-equal? t2 Err)) A
+   #:return-when (type-equal? t1 -Bottom) A
+   (define mask1 (Type-mask t1))
+   (define mask2 (Type-mask t2))
+   #:return-when (disjoint-masks? mask1 mask2) #f
+   #:return-when (type-equal? t1 t2) A
+   (define t1-subtype-cache (Type-subtype-cache t1))
+   (define cr (hash-ref t1-subtype-cache (Rep-seq t2) 'missing))
    #:return-when (boolean? cr) (and cr A)
-   (define ks (Type-key s))
-   (define kt (Type-key t))
-   #:return-when (and (symbol? ks) (symbol? kt) (not (eq? ks kt))) #f
-   #:return-when (and (symbol? ks) (pair? kt) (not (memq ks kt))) #f
-   #:return-when 
-   (and (pair? ks) (pair? kt)
-        (for/and ([i (in-list ks)]) (not (memq i kt))))
-   #f
-   #:return-when (eq? ss bottom-key) A
-   #:return-when (eq? st top-key) A
-   (define A0 (remember s t A))
-   (define r
-     ;; FIXME -- make this go into only the places that need it -- slows down new-metrics.rkt significantly
-     (update-current-seen A0
-       (match* (s t)
-         ;; these cases are above as special cases
-         ;; [((Union: (list)) _) A0] ;; this is extremely common, so it goes first
-         ;; [(_ (Univ:)) A0]
-         ;; error is top and bot
```

Comment:
equality testing on Unions can be much more costly than checking type-masks, so my vote is it stays where it is.